GH-3785: Close stream for persistent collection

artembilan opened the pull request:
Fixes https://github.com/spring-projects/spring-integration/issues/3785

* Fix `CollectionArgumentResolver` and `PayloadsArgumentResolver` to
close the `Stream` of message after its usage
* Rework `AbstractKeyValueMessageStore.removeMessagesFromGroup()`
to iterate input collection of messages not its stream to avoid
the mentioned problem

**Cherry-pick to `5.5.x`**

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

artembilan committed:
GH-3785: Close stream for persistent collection

Fixes https://github.com/spring-projects/spring-integration/issues/3785

* Fix `CollectionArgumentResolver` and `PayloadsArgumentResolver` to
close the `Stream` of message after its usage
* Rework `AbstractKeyValueMessageStore.removeMessagesFromGroup()`
to iterate input collection of messages not its stream to avoid
the mentioned problem

**Cherry-pick to `5.5.x`**

garyrussell commented:
No tests needed?

artembilan commented:
Well, it doesn't look obvious what test has to do, but I'll think about that tomorrow.
For self-learning purpose.
I just don't want to add more deps for no extreme reason....

garyrussell commented:
Maybe a mock AMGS with a message stream passed into each of the resolvers and verify that the mock stream is closed?

Perhaps we also should enhance the javadocs of the JDBC MGS to say that the stream must be closed; the AGMS just streams the `getMessages()` so close is not required there, but any subclass that uses true streaming presumably needs the stream closed.

It looks like we don't even have any unit tests for `PMG.streamMessages()`.

artembilan committed:
* Add `JdbcMessageStoreTests.testMessageGroupStreamNoConnectionPoolLeak()`
to ensure that no leaks in the connection pool anymore.
* Improve `MessageGroupStore.streamMessagesForGroup()` JavaDocs about
requirements to close the `Stream` from persistent message store impls

garyrussell commented:
... and cherry-picked.
